numpy/__init__.pyi

```diff
@@ -358,6 +358,187 @@ class ndarray(Iterable, Sized, SupportsInt, SupportsFloat, SupportsComplex,
+        def __hex__(self) -> str: ...
+        def __nonzero__(self) -> bool: ...
+        def __unicode__(self) -> Text: ...
+    else:
+        def __bool__(self) -> bool: ...
+        def __bytes__(self) -> bytes: ...
+    def __str__(self) -> str: ...
+    def __repr__(self) -> str: ...
+
+    def __copy__(self: _Scalar, order: str = ...) -> _Scalar: ...
+    def __deepcopy__(self: _Scalar, memo: dict) -> _Scalar: ...
+
+    def __lt__(self, other): ...
+    def __le__(self, other): ...
+    def __eq__(self, other): ...
+    def __ne__(self, other): ...
+    def __gt__(self, other): ...
+    def __ge__(self, other): ...
+
+    def __add__(self, other): ...
+    def __radd__(self, other): ...
+    def __iadd__(self, other): ...
+
+    def __sub__(self, other): ...
+    def __rsub__(self, other): ...
+    def __isub__(self, other): ...
+
+    def __mul__(self, other): ...
+    def __rmul__(self, other): ...
+    def __imul__(self, other): ...
+
+    if sys.version_info[0] < 3:
+        def __div__(self, other): ...
+        def __rdiv__(self, other): ...
+        def __idiv__(self, other): ...
+
+    def __truediv__(self, other): ...
+    def __rtruediv__(self, other): ...
+    def __itruediv__(self, other): ...
+
+    def __floordiv__(self, other): ...
+    def __rfloordiv__(self, other): ...
+    def __ifloordiv__(self, other): ...
+
+    def __mod__(self, other): ...
+    def __rmod__(self, other): ...
+    def __imod__(self, other): ...
+
+    def __divmod__(self, other): ...
+    def __rdivmod__(self, other): ...
+
+    # NumPy's __pow__ doesn't handle a third argument
+    def __pow__(self, other): ...
+    def __rpow__(self, other): ...
+    def __ipow__(self, other): ...
+
+    def __lshift__(self, other): ...
+    def __rlshift__(self, other): ...
+    def __ilshift__(self, other): ...
+
+    def __rshift__(self, other): ...
+    def __rrshift__(self, other): ...
+    def __irshift__(self, other): ...
+
+    def __and__(self, other): ...
+    def __rand__(self, other): ...
+    def __iand__(self, other): ...
+
+    def __xor__(self, other): ...
+    def __rxor__(self, other): ...
+    def __ixor__(self, other): ...
+
+    def __or__(self, other): ...
+    def __ror__(self, other): ...
+    def __ior__(self, other): ...
+
+    if sys.version_info[:2] >= (3, 5):
+        def __matmul__(self, other): ...
+        def __rmatmul__(self, other): ...
+
+    def __neg__(self) -> ndarray: ...
+    def __pos__(self) -> ndarray: ...
+    def __abs__(self) -> ndarray: ...
+    def __invert__(self) -> ndarray: ...
+
+    # TODO(shoyer): remove when all methods are defined
+    def __getattr__(self, name) -> Any: ...
+
+class number(generic): ...
+class bool_(generic): ...
+class object_(generic): ...
+class datetime64(generic): ...
+
+class integer(number): ...
+class signedinteger(integer): ...
+class int8(signedinteger): ...
+class int16(signedinteger): ...
+class int32(signedinteger): ...
+class int64(signedinteger): ...
```

Comment:
Sadly this doesn't actually reflect the underlying model - in reality, the types are `byte`, `short`, `int`, `long`, and `longlong`, and the numeric aliases are attached in a platform-dependent manner.

Reply:
Does it really matter whether we define aliases or the real classes here? I would actually be OK with doing only these aliases. For annotations (and clearly written code in general), it's best to avoid platform dependent types like `int`.

Reply:
Well since there are actually 5 underlying types, sometimes numpy returns a type that is not referred to by any of these four aliases (which is made worse by numpy/numpy#10151 making them look the same).

I think what you have there is an argument about removing the unsized types from numpy - which I think is valid. But I think the type annotations should reflect what actually happens, not how we want things to happen.